packages/protocol/src/utils/SPID.ts

```diff
@@ -0,0 +1,202 @@
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new SPIDValidationError('SPID validation failed: id must be non-empty string: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new SPIDValidationError('SPID validation failed: partition must be a safe integer >= 0: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDLike): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = SPID.from(other)
+        } catch (_err) {
+            // ignore error
+            // not equal if not valid
+            return false
+        }
+
+        // check key matches
+        return this.key === otherSpid.key
+    }
+
+    /**
+     * Convert SPIDLikePartial to SPIDObjectPartial
+     * i.e. normalizes various input types/shapes to { streamId?, streamPartition? }
+     * Note: does not throw on malformed input
+     */
+    static toSPIDObjectPartial(spidLike: SPIDLikePartial): SPIDObjectPartial {
+        // convert from string
+        if (typeof spidLike === 'string') {
+            const [streamId, partitionStr] = spidLike.split(this.SEPARATOR)
+            // partition is optional, falls back to undefined i.e. default
+            const streamPartition = partitionStr != null ? Number.parseFloat(partitionStr) : undefined
+            return { streamId, streamPartition }
+        } else if (spidLike && typeof spidLike === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidLike.streamId || spidLike.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const partition = spidLike.streamPartition || spidLike.partition
+            // try parse if a value was passed, but fall back to undefined i.e. default
+            const streamPartition = partition != null ? Number.parseFloat(partition) : undefined
+            return { streamId, streamPartition }
+        } else {
+            return { streamId: undefined, streamPartition: undefined }
+        }
+    }
+
+    /**
+     * Convert to SPID if possible, with defaults.
+     * e.g.
+     * ```ts
+     * fromDefaults(streamId, { partition: 0 })
+     * ```
+     */
+    static fromDefaults(spidLike: SPIDLike, defaultValues?: SPIDLikePartial): SPID
+    static fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial): SPID // requires id+partition if no defaults
+    static fromDefaults(spidLike: SPIDLikePartial, defaultValues?: SPIDLikePartial): SPID {
```

Comment:
Wouldn't having only the third declaration be enough here since type `SPIDLikePartial` already includes `SPIDLike` as one of its options?

Reply:
Right, but that will always be the case in TS, the last overload always has to match all the previous ones.

In TS method/function overloads:
* select the first matching signature
* and the last "implementation" signature always has to be a superset of all the other signatures.

e.g. you can't do something like this
```typescript
class A {
  convert(v: string)
  convert(v: number) { // This overload signature is not compatible with its implementation signature.
      
  } 
}
```
----

In our case, we want TS to error when you give a partial input for the spid, and also nothing for a default, which we can detect is wrong at type compile time:

```js
// partial input, no defaults
fromDefaults({ streamId }) // TypeScript error (wants a `SPIDLike`, not a `SPIDLikePartial`)
```

The first overload matches a non-partial, and keeps the second argument optional. This has to go first since the second overload would also match this signature and thus be selected, and would require default values even when you don't input a partial object.

```typescript
static fromDefaults(spidLike: SPIDLike, defaultValues?: SPIDLikePartial): SPID
```

The second overload tells TS to *require* `defaultValues` to be passed if a partial SPID was passed (or something like it) i.e. `SPIDLikePartitial`. This is the bit we want.
```typescript
static fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial): SPID // requires id+partition if no defaults
```

The third overload is the implementation signature and should match all prior signatures, and acts as a catch all:
```typescript
static fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial): SPID // requires id+partition if no defaults
```

Reply:
I think I'm on to what you're saying, one question though.

Let's say I pass as 1st argument an instance of `SPIDLikePartial` and 2nd argument is left `undefined`. The first overload is skipped  since the 1st argument type doesn't "match" (can't assign `SPIDLikePartial` to `SPIDLike`). We are left thus with
```ts
static fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial): SPID // requires id+partition if no defaults
static fromDefaults(spidLike: SPIDLikePartial, defaultValues?: SPIDLikePartial): SPID {
```

Is it then so that the given arguments must match _both_ of these overloaded signatures, or is an _or_ style ordeal? I.e. it fails the `fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial)` "match" (because 2nd argument is undefined), but will "match" the last, catch-all overload? So I'm assuming it is not enough for it to just match the catch-all signature?

Reply:
Ahh or is more like that there like a special interpretation of the last signature when giving multiple? Let's say we have some method with abstract arguments:
```ts
static foo(...sig1)
static foo(...sig2)
static foo(...sig3)
static foo(...sig4) {
  ....
}
```

Is there like a special rule that you can't use `static foo(...sig4)` directly to call method, instead you must use one of the above options? But the limiting thing is that the other options must be compatible with `static foo(...sig4)` ?